File: 940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int right=0;
        int left=0;
        int maxi=0;
        map<int,int>mp;
        while(right<fruits.size()){
            mp[fruits[right]]++;
            while(mp.size()>2){
                mp[fruits[left]]--;
                if(mp[fruits[left]]==0){
                    mp.erase(fruits[left]);
                }
                left++;
            }
            maxi=max(maxi,right-left+1);
            right++;
        }
        return maxi;
        
    }
};

File: 940-fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        HashMap<Integer,Integer>mp=new HashMap<>();
        int left=0;
        int right=0;
        int maxi=Integer.MIN_VALUE;
        while(right<fruits.length){
            mp.put(fruits[right],mp.getOrDefault(fruits[right],0)+1);
            while(mp.size()>2){
                mp.put(fruits[left],mp.getOrDefault(fruits[left],0)-1);
                if(mp.get(fruits[left])==0)mp.remove(fruits[left]);
                left++;
            }
            maxi=Math.max(maxi,right-left+1);
            right++;

        }
        return maxi;

        
    }
}